qutebrowser/browser/tab.py

```
@@ -0,0 +1,525 @@
+
+        for properties, see WebView/WebEngineView docs.
+
+    Signals:
+        See related Qt signals.
+
+        new_tab_requested: Emitted when a new tab should be opened with the
+                           given URL.
+    """
+
+    window_close_requested = pyqtSignal()
+    link_hovered = pyqtSignal(str)
+    load_started = pyqtSignal()
+    load_progress = pyqtSignal(int)
+    load_finished = pyqtSignal(bool)
+    icon_changed = pyqtSignal(QIcon)
+    # FIXME:qtwebengine get rid of this altogether?
+    url_text_changed = pyqtSignal(str)
+    title_changed = pyqtSignal(str)
+    load_status_changed = pyqtSignal(str)
+    new_tab_requested = pyqtSignal(QUrl)
+    shutting_down = pyqtSignal()
+
+    def __init__(self, win_id, parent=None):
+        self.win_id = win_id
+        self.tab_id = next(tab_id_gen)
+        super().__init__(parent)
+        # self.history = AbstractHistory(self)
+        # self.scroll = AbstractScroller(parent=self)
+        # self.caret = AbstractCaret(win_id=win_id, tab=self, modeman=...,
+        #                            parent=self)
+        # self.zoom = AbstractZoom(win_id=win_id)
+        # self.search = AbstractSearch(parent=self)
+        self.data = TabData()
+        self._layout = None
+        self._widget = None
+        self.backend = None
+
+    def _set_widget(self, widget):
+        # pylint: disable=protected-access
+        self._layout = WrapperLayout(widget, self)
+        self._widget = widget
+        self.history._history = widget.history()
+        self.scroll._widget = widget
+        self.caret._widget = widget
+        self.zoom._widget = widget
+        self.search._widget = widget
+        widget.mouse_wheel_zoom.connect(self.zoom._on_mouse_wheel_zoom)
+        widget.setParent(self)
+        self.setFocusProxy(widget)
+
+    @pyqtSlot()
+    def _on_load_started(self):
+        self.data.viewing_source = False
+        self.load_started.emit()
+
+    @property
+    def cur_url(self):
```

Are you confident that those functions wont raise exceptions? If not, a normal function would probably be better, since PyQt might hide the exceptions otherwise.